### PR TITLE
Created hydrator strategy chain

### PR DESCRIPTION
**Pull request by ojhaujjwal**
Feedbacks are welcome!

### CODE REVIEW

File: library/Zend/Stdlib/Hydrator/Strategy/Exception/DomainException.php [hunk 1]
```diff
@@ -0,0 +1,14 @@
+<?php
+/**
+ * Zend Framework (http://framework.zend.com/)
+ *
+ * @link      http://github.com/zendframework/zf2 for the canonical source repository
+ * @copyright Copyright (c) 2005-2014 Zend Technologies USA Inc. (http://www.zend.com)
```
Comment (samsonasik) [inline library/Zend/Stdlib/Hydrator/Strategy/Exception/DomainException.php]: 2015 ;)

File: library/Zend/Stdlib/Hydrator/Strategy/Exception/ExceptionInterface.php [hunk 2]
```diff
@@ -0,0 +1,14 @@
+<?php
+/**
+ * Zend Framework (http://framework.zend.com/)
+ *
+ * @link      http://github.com/zendframework/zf2 for the canonical source repository
+ * @copyright Copyright (c) 2005-2014 Zend Technologies USA Inc. (http://www.zend.com)
```
Comment (samsonasik) [inline library/Zend/Stdlib/Hydrator/Strategy/Exception/ExceptionInterface.php]: 2015 ;)

File: tests/ZendTest/Stdlib/Hydrator/Strategy/StrategyChainTest.php [hunk 3]
```diff
@@ -0,0 +1,142 @@
+<?php
+/**
+ * Zend Framework (http://framework.zend.com/)
+ *
+ * @link      http://github.com/zendframework/zf2 for the canonical source repository
+ * @copyright Copyright (c) 2005-2014 Zend Technologies USA Inc. (http://www.zend.com)
```
Comment (samsonasik) [inline tests/ZendTest/Stdlib/Hydrator/Strategy/StrategyChainTest.php]: 2015 ;)

File: library/Zend/Stdlib/Hydrator/Strategy/StrategyChain.php [hunk 4]
```diff
@@ -0,0 +1,127 @@
+<?php
+/**
+ * Zend Framework (http://framework.zend.com/)
+ *
+ * @link      http://github.com/zendframework/zf2 for the canonical source repository
+ * @copyright Copyright (c) 2005-2014 Zend Technologies USA Inc. (http://www.zend.com)
```
Comment (samsonasik) [inline library/Zend/Stdlib/Hydrator/Strategy/StrategyChain.php]: 2015 ;)

File: library/Zend/Stdlib/Hydrator/Strategy/Exception/InvalidArgumentException.php [hunk 5]
```diff
@@ -0,0 +1,14 @@
+<?php
+/**
+ * Zend Framework (http://framework.zend.com/)
+ *
+ * @link      http://github.com/zendframework/zf2 for the canonical source repository
+ * @copyright Copyright (c) 2005-2014 Zend Technologies USA Inc. (http://www.zend.com)
```
Comment (samsonasik) [inline library/Zend/Stdlib/Hydrator/Strategy/Exception/InvalidArgumentException.php]: 2015 ;)

File: library/Zend/Stdlib/Hydrator/Strategy/StrategyChain.php (hunk 4 above)
Comment (Ocramius) [inline library/Zend/Stdlib/Hydrator/Strategy/StrategyChain.php]: @samsonasik I'd rather change these on merge instead of poking each author of each PR.

**Comment by ojhaujjwal** [comment]
I have made this class final and removed unnecessary setters/getters!

File: library/Zend/Stdlib/Hydrator/Strategy/StrategyChain.php [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+
+        return $value;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function extract($value)
+    {
+        foreach (array_reverse(iterator_to_array($this->strategies)) as $strategy) {
```
Comment (ojhaujjwal) [inline library/Zend/Stdlib/Hydrator/Strategy/StrategyChain.php]: I am not sure how to do that! :/

Comment (Ocramius) [inline library/Zend/Stdlib/Hydrator/Strategy/StrategyChain.php]: @ojhaujjwal in the constructor :-)

File: library/Zend/Stdlib/Hydrator/Strategy/StrategyChain.php [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+    /**
+     * Sets strategies
+     *
+     * @param  array|Traversable                  $strategies
+     * @return void
+     * @throws Exception\InvalidArgumentException
+     */
+    private function setStrategies($strategies)
+    {
+        if (!is_array($strategies) && !$strategies instanceof Traversable) {
```
Comment (Ocramius) [inline library/Zend/Stdlib/Hydrator/Strategy/StrategyChain.php]: You can probably skip all this and use `ArrayUtils::iteratorToArray()` instead, which already encapsulates this conditional

File: library/Zend/Stdlib/Hydrator/Strategy/StrategyChain.php [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+            ));
+        }
+
+        foreach ($strategies as $value) {
+            if (is_array($value)) {
+                if (!isset($value['strategy'])) {
+                    throw new Exception\DomainException('No strategy is provided.');
+                }
+                $strategy = $value['strategy'];
+                $priority = isset($value['priority']) ? $value['priority'] : self::DEFAULT_PRIORITY;
```
Comment (Ocramius) [inline library/Zend/Stdlib/Hydrator/Strategy/StrategyChain.php]: Do we really need to handle priorities? What if the order of the given `$strategies` (in the constructor) was enough instead?  That would simplify logic greatly

File: tests/ZendTest/Stdlib/Hydrator/Strategy/StrategyChainTest.php [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+ * @copyright Copyright (c) 2005-2015 Zend Technologies USA Inc. (http://www.zend.com)
+ * @license   http://framework.zend.com/license/new-bsd New BSD License
+ */
+
+namespace ZendTest\Stdlib\Hydrator\Strategy;
+
+use Zend\Stdlib\Hydrator\Strategy\StrategyChain;
+use Zend\Stdlib\Hydrator\Strategy\ClosureStrategy;
+
+class StrategyChainTest extends \PHPUnit_Framework_TestCase
```
Comment (Ocramius) [inline tests/ZendTest/Stdlib/Hydrator/Strategy/StrategyChainTest.php]: Please add coverage annotations here

File: library/Zend/Stdlib/Hydrator/Strategy/StrategyChain.php (hunk 8 above)
Comment (Ocramius) [inline library/Zend/Stdlib/Hydrator/Strategy/StrategyChain.php]: Note that priorities can still be handled by a static constructor later on (if needed)

**Comment by ojhaujjwal** [comment]
Okay. I have removed priorities support!

**Comment by weierophinney** [comment]
Merged to develop for release with 2.4.0.

@ojhaujjwal Please write up some documentation for the https://github.com/zendframework/zf2-documentation repository! :)

**Pull request closed without merging**